apps/aestratum/test/aestratum_client_suite_utils.erl

```diff
@@ -0,0 +1,257 @@
+          <<"worker">> => <<"worker1">>},
+      <<"miners">> =>
+        [#{<<"exec">> => <<"mean29-generic">>,
+           <<"exec_group">> => <<"aecuckoo">>,
+           <<"extra_args">> => <<"">>,
+           <<"hex_enc_hdr">> => false,
+           <<"repeats">> => 100,
+           <<"edge_bits">> => 29}]
+     }.
+
+write_config(F, _CfgSchema, Cfg) ->
+    JSON = jsx:prettify(jsx:encode(Cfg)),
+    ok = file:write_file(F, JSON),
+    ct:log("Writing config (~p)~n~s", [F, JSON]),
+    ok = file:write_file(F, JSON).
+
+make_stratum_clients(Cfg, NodesList) ->
+    TopClientDir = ?config(top_client_dir, Cfg),
+    ct:log("Top client dir = ~p", [TopClientDir]),
+    Client =
+        filename:join(TopClientDir,
+                      "_build/test/rel/aestratum_client"),
+    [setup_node(N, TopClientDir, Client, Cfg) || N <- NodesList].
+
+setup_node(N, Top, Client, Cfg) ->
+    ct:log("setup_node(~p,Cfg)", [N]),
+    DDir = node_shortcut(N, Cfg),
+    filelib:ensure_dir(filename:join(DDir, "foo")),
+    cp_dir(filename:join(Client, "releases"), DDir ++ "/"),
+    cp_dir(filename:join(Client, "bin"), DDir ++ "/"),
+    symlink(filename:join(Client, "lib"), filename:join(DDir, "lib")),
+    {ok, VerContents} = file:read_file(filename:join(Client, "VERSION")),
+    [VerB |_ ] = binary:split(VerContents, [<<"\n">>, <<"\r">>], [global]),
+    Version = binary_to_list(VerB),
+    %%
+    CfgD = filename:join([Top, "config/", "test"]),
+    RelD = filename:dirname(filename:join([DDir, "releases", Version, "aestratum_client.rel"])),
+    cp_file(filename:join(CfgD, "sys.config"),
+            filename:join(RelD, "sys.config")),
+    cp_file(filename:join(CfgD, "vm.args"),
+            filename:join(RelD, "vm.args")),
+    delete_file(filename:join(RelD, "vm.args")), %% cookie and node name
+    make_vm_args_file(filename:join(RelD, "vm.args"), N),
+    delete_file(filename:join(RelD, "vm.args.orig")),
+    delete_file(filename:join(RelD, "sys.config.orig")).
+
+connect(N, Timeout) when Timeout < 10000 ->
+    timer:sleep(Timeout),
+    case net_kernel:hidden_connect(N) of
+        true ->
+            ct:log("hidden_connect(~p) -> true", [N]),
+            await_status(N, 5),
+            true;
+        false ->
+            ct:log("hidden_connect(~p) -> false, retrying ...", [N]),
+            connect(N, Timeout * 2)
+    end;
+connect(N, _) ->
+    ct:log("exhausted retries (~p)", [N]),
+    erlang:error({could_not_connect, N}).
+
+report_node_config(N) ->
+    [ct:log("~w env: ~p", [A, rpc:call(N, application, get_all_env, [A], 2000)]) ||
+        A <- [aestratum_client]].
+
+await_status(N, Retries) when Retries > 0 ->
+    case client_status(N) of
+        S when is_map(S) ->
+            ok;
+        _Other ->
+            timer:sleep(1000),
+            await_status(N, Retries - 1)
+    end;
+await_status(_N, 0) ->
+    error(timeout_waiting_for_status).
+
+client_status(N) ->
+    rpc:call(N, aestratum_client, status, []).
```

Comment:
This function call makes the `aestratum_client_handler` process [call the `aestratum_client_session:status` function even if `Session` undefined](https://github.com/aeternity/aestratum_client/blob/v0.1/src/aestratum_client_handler.erl#L137) killing the `aestratum_client_handler`. The symptom I observed locally is the `await_status` timed out while the integration test called `aestratum_client_suite_utils:connect(C1Node)`.

This brings to some questions:
* What is value of `Session` meant to be at that point in the test, and can it be `undefined` for timing?
* If it can be `undefined`, shall `aestratum_client_handler` process die? Shall test code here or client code be amended.

Reply:
It's here: https://github.com/aeternity/aestratum_client/blob/master/src/aestratum_client_session.erl#L117

So it needs to be a valid #session record.